### PR TITLE
[countersyncd]: resolve object names by template id

### PR DESCRIPTION
**Pull request by Pterosaur**
**What I did**

Fixed `countersyncd` IPFIX object-name resolution to use the current data set's `template_id` when selecting `object_names`, instead of picking an arbitrary key from the temporary/applied template maps.

Added a regression unit test covering two template IDs with different `object_names` in the same input stream.

**Why I did it**

The previous logic could associate a data record with the wrong `object_names` set when multiple templates or sessions coexisted. That is a correctness issue: counters can be attributed to the wrong object instead of just being imprecise.

**How I verified it**

- Ran targeted regression test in a CI-like container environment:
  - `cargo test --locked --manifest-path crates/countersyncd/Cargo.toml test_object_names_follow_template_id -- --nocapture`
- Confirmed the targeted test passed.
- Checked compile output during test build and did not observe new Rust compiler warnings from this change.

**Details if related**

### CODE REVIEW

File: crates/countersyncd/src/actor/ipfix.rs [hunk 1]
```diff
@@ -584,6 +584,15 @@ impl IpfixActor {
...
     }
 
+    fn get_template_key(&self, template_id: u16) -> Option<&String> {
+        self.temporary_templates_map.get(&template_id).or_else(|| {
+            self.applied_templates_map
+                .iter()
+                .find(|(_, template_ids)| template_ids.contains(&template_id))
+                .map(|(msg_key, _)| msg_key)
+        })
+    }
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply crates/countersyncd/src/actor/ipfix.rs:587]: A `template_id` can never exist in both maps simultaneously. `update_applied_template()` uses `retain()` to remove all matching entries from `temporary_templates_map` before inserting into `applied_templates_map`, so the transition is atomic within a single-threaded context. The `or_else` fallback is safe by design.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 2]
```diff
@@ -1167,6 +1165,68 @@ mod test {
...
+
+        let all_object_names: Vec<&str> = stats
+            .iter()
+            .flat_map(|msg| msg.stats.iter().map(|s| s.object_name.as_str()))
+            .collect();
+
+        assert!(all_object_names.contains(&"Ethernet0"));
+        assert!(all_object_names.contains(&"Ethernet1"));
+        assert!(all_object_names.contains(&"Ethernet8"));
+        assert!(all_object_names.contains(&"Ethernet12"));
```
[the remark this answers is not in the document]
Comment (Pterosaur) [reply crates/countersyncd/src/actor/ipfix.rs]: The test already enforces no cross-contamination via the `all()` + `||` pattern on lines 1232-1238: each record must contain names exclusively from session A or session B. If any record mixed names from both sessions, the assertion would fail. This is functionally equivalent to the suggested grouping approach.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 3]
```diff
@@ -584,6 +584,15 @@ impl IpfixActor {
         self.applied_templates_map.insert(msg_key, template_ids);
     }
 
+    fn get_template_key(&self, template_id: u16) -> Option<&String> {
+        self.temporary_templates_map.get(&template_id).or_else(|| {
+            self.applied_templates_map
+                .iter()
+                .find(|(_, template_ids)| template_ids.contains(&template_id))
+                .map(|(msg_key, _)| msg_key)
```
Comment (banidoru) [inline crates/countersyncd/src/actor/ipfix.rs:585]: Nit: This linear scan of `applied_templates_map` is O(n × m) where n = number of session keys and m = average template IDs per key. Fine for typical workloads, but if the number of sessions grows significantly, consider building a reverse index (`HashMap<u16, String>`) that maps template_id → session key directly, updated alongside `applied_templates_map` in `update_applied_template()`.

**Review by banidoru — APPROVED** [review]
Good bug fix. The core change — resolving object names per-template via the data set's `set_id` instead of grabbing the first arbitrary key from the template maps — is correct and well-motivated.

Key observations:
- `get_template_key` correctly prioritizes the temporary map and falls back to a linear scan of `applied_templates_map`. The linear scan is acceptable given expected map sizes but worth noting.
- Clearing stale `object_names_map` entries when `object_names` is `None` prevents silent data corruption — good defensive fix.
- The restructured `handle_record` loop over `data_message.sets` is cleaner and more correct than the previous flat iteration.
- Both new tests cover the critical scenarios (multi-session template isolation, stale mapping cleanup).
- The updated test (`test_object_names_follow_template_id`) properly asserts no cross-contamination via the `only_session_a || only_session_b` pattern with `all()` checks.

One minor inline comment below.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 4]
```diff
@@ -777,126 +790,115 @@ impl IpfixActor {
...
+            for set in &data_message.sets {
+                let (template_id, datarecords) = match &set.records {
+                    ipfixrw::parser::Records::Data { set_id, data } => (*set_id, data),
+                    _ => continue,
+                };
+
+                let object_names = self
+                    .get_template_key(template_id)
+                    .and_then(|key| self.object_names_map.get(key))
+                    .map(|names| names.as_slice())
```
Comment (banidoru) [inline crates/countersyncd/src/actor/ipfix.rs]: Good design choice: resolving `object_names` once per set (outside the record loop) is both correct and avoids redundant map lookups. Since all records within a single IPFIX data set share the same `template_id`, this is the right granularity.

File: crates/countersyncd/src/actor/ipfix.rs [hunk 5]
```diff
@@ -621,10 +630,14 @@ impl IpfixActor {
             }
         }
 
-        // Store object names if provided
+        // Update object name mapping for the session key.
+        // A missing object_names field means the latest template update no longer
+        // provides object name mapping, so any stale value must be cleared.
         if let Some(object_names) = &templates.object_names {
             self.object_names_map
```
Comment (banidoru) [inline crates/countersyncd/src/actor/ipfix.rs]: This `else` branch is an important correctness fix. Without it, a template update that no longer carries object names would leave a stale mapping in place, causing records to silently get wrong object names. Good catch.

**Review by banidoru — APPROVED** [review]
This is a well-motivated fix for a real correctness bug — the old code grabbed the first arbitrary key from the template maps, making object name assignment effectively random when multiple sessions existed. The new approach correctly resolves object names per data set via template_id lookup.

Key observations:
- `get_template_key()` correctly prioritizes `temporary_templates_map` (pre-apply) then falls back to `applied_templates_map`.
- Clearing stale `object_names_map` entries when `object_names` is `None` prevents silent data corruption from outdated mappings.
- Moving object name resolution outside the per-record inner loop (once per set) is both correct and more efficient.
- The updated test (`test_object_names_follow_template_id`) now properly asserts no cross-contamination between sessions.
- The `test_template_update_without_object_names_clears_stale_mapping` test directly validates the new cleanup behavior.

The linear scan in `get_template_key` (noted by another reviewer) is the only minor concern but is acceptable for expected workload sizes. LGTM.

**Review by banidoru — APPROVED** [review]
Solid fix. The core bug was that the old code grabbed an arbitrary session key from the template maps, causing object names to be cross-contaminated across sessions. The new approach correctly resolves object names per data set via template_id → session key lookup.

Key observations:
- **Correctness**: Iterating `data_message.sets` and extracting `set_id` (template_id) per set is the right approach — all records in an IPFIX data set share the same template, so resolving object names once per set is both correct and efficient.
- **Stale mapping cleanup**: The `else` branch removing `object_names_map` entries when `object_names` is `None` prevents a real data-corruption scenario. Good defensive fix.
- **get_template_key**: Clean helper. The linear scan of `applied_templates_map` is fine for expected workloads (already noted by other reviewers).
- **Tests**: Both new tests cover the important cases (multi-session isolation and stale-clearing). The updated `test_object_names_follow_template_id` with per-record exclusive assertions is stronger than simple contains checks.
- **Code reduction**: ~30 lines of broken template-key heuristic replaced with a clean 8-line helper. Net improvement in readability.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

File: crates/countersyncd/src/actor/ipfix.rs (hunk 3 above)
Comment (Pterosaur) [reply crates/countersyncd/src/actor/ipfix.rs:585]: Keeping the current shape for now. The extra reverse index is a reasonable future optimization, but the current lookup happens once per IPFIX data set (not per field anymore) and the expected session cardinality here is still small enough that I do not want to add another structure to keep in sync in this bug-fix PR.

**Comment by Pterosaur** [comment]
Created 202412 backport PR: https://github.com/Azure/sonic-swss.msft/pull/211

**Review by r12f — DISMISSED** [review]

**Review by banidoru — APPROVED** [review]
- New commit is merge-from-master only — no changes to PR code
- All prior review threads already resolved
- No new issues found

**Review by banidoru — APPROVED** [review]
- Correct and efficient object name resolution.
- Stale mapping cleanup is a good correctness fix.
- Template transition logic is confirmed to be safe.
- Existing test coverage is sufficient.

**Review by banidoru — APPROVED** [review]
- **Approve.** Core bug fix is correct — object names now resolve by template_id via `get_template_key()` instead of grabbing the first arbitrary map entry.
- Per-set iteration in `handle_record` is the right granularity; avoids redundant lookups and matches the IPFIX data model.
- Stale `object_names_map` cleanup on template updates without object names prevents silent wrong-name assignment.
- Both new tests are well-structured: cross-contamination test validates isolation, stale-mapping test validates cleanup.
- All prior review threads addressed and resolved.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
- Object name resolution correctly scoped per template_id via `get_template_key()` — fixes the cross-contamination bug
- Stale object_names properly cleared when template update omits them
- Set-level iteration with single `object_names` lookup per set is clean and efficient
- Both new tests (`test_object_names_follow_template_id`, `test_template_update_without_object_names_clears_stale_mapping`) cover the core fix and the edge case well
- No new issues found in iteration 3

**Review by banidoru — APPROVED** [review]
- **Approve** — all prior concerns addressed, code is solid
- `get_template_key()` correctly resolves template_id via both maps with safe `or_else` fallback
- Stale object_names cleanup on template updates without names prevents silent data corruption
- Set-level object_names resolution eliminates redundant per-field lookups — clean and correct
- Both new tests cover the core bug (cross-contamination) and the stale-mapping edge case well

**Review by banidoru — APPROVED** [review]
- No code changes since last review — only a merge-from-master commit
- All 5 prior review threads resolved with satisfactory responses
- Core fix remains correct: template_id-based object name resolution eliminates cross-contamination
- Test coverage adequate for both the fix and stale-mapping cleanup
- LGTM

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
- Core fix is correct: object_names now resolved per data set via template_id, eliminating cross-contamination
- Stale mapping cleanup (else branch clearing object_names_map) prevents silent wrong-name attribution
- get_template_key() lookup logic is sound; temporary→applied fallback is safe given mutual exclusivity
- Both new tests are well-structured: cross-contamination test and stale-clearing test cover the key scenarios
- Code is cleaner post-refactor — set-level iteration is more readable than the old per-field template_key hack

**Review by banidoru — APPROVED** [review]
- **Overall:** The changes are a significant improvement. The test refactoring enhances code quality and maintainability by properly encapsulating test setup and verification logic within the respective wrapper classes.
- **New Features:** The newly added tests for L2/L3 multicast and mirror sessions provide valuable coverage for these features.
- **Testing:** The core logic of the tests remains sound, and the refactoring correctly moves the responsibility of response verification and cleanup to the test object wrappers.

**Review by banidoru — APPROVED** [review]
- No code changes since last review — latest commit is merge-from-master only
- All prior review threads resolved
- Core fix remains correct: per-set template_id lookup eliminates object name cross-contamination
- Stale mapping cleanup and both regression tests are solid
- LGTM

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
$Clean iteration. All prior feedback addressed.\n\n- `get_template_key()` extraction is well-structured\n- Per-set object name resolution via `template_id` correctly fixes the cross-contamination bug\n- Stale mapping cleanup (`else { remove }`) is a good defensive addition\n- Both new tests are well-designed: cross-session isolation and stale-clear coverage\n- No new issues found

**Review by banidoru — APPROVED** [review]
- Template-to-object-name lookup via `get_template_key` correctly resolves per data set, eliminating cross-contamination
- Stale `object_names_map` cleanup on `None` is an important correctness fix
- Object names resolved once per set (not per field) — good performance improvement
- Both new tests cover the key scenarios well; the `all()` + `||` pattern catches cross-contamination
- All prior review concerns addressed and resolved

**Review by banidoru — APPROVED** [review]
- Object-name resolution correctly moved from per-field (with wrong arbitrary template lookup) to per-set using actual template ID from set header
- Stale object_names cleanup on template update without names is a solid correctness fix
- `get_template_key` linear scan is fine for expected cardinality; reverse index noted as future optimization
- Both new tests are well-structured and cover the key scenarios (multi-session isolation, stale clearing)
- Clean simplification overall — removed ~40 lines of broken template-key guessing logic

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Review by banidoru — APPROVED** [review]
- CRM decrement now correctly guarded behind `SAI_STATUS_SUCCESS` only — fixes over-decrement on handled errors like `ITEM_NOT_FOUND`
- Early-return refactor in `removeOutboundRoutingPost` / `removeInboundRoutingPost` improves readability and eliminates fall-through risk
- Good test coverage: `RemoveNonexistOutboundRoutingDoesNotDecrementCrm` and inbound variant verify the fix
- `handleSaiRemoveStatus` correctly extended to cover DASH routing APIs
- Minor: trailing whitespace on `mock_dash_orch_test.cpp:104` and missing newline at EOF in `dashrouteorch_ut.cpp` — non-blocking

**Verdict: APPROVE**

**Review by banidoru — APPROVED** [review]
- **Approved.** Core fix is correct: object names now resolve per data set via `template_id`, eliminating cross-session contamination from the old "grab first key" approach.
- `get_template_key()` cleanly encapsulates the two-map lookup with proper short-circuit semantics; mutual exclusivity between `temporary_templates_map` and `applied_templates_map` is maintained by `update_applied_template()`.
- The `else` branch clearing stale `object_names_map` entries is an important correctness fix — prevents silent wrong-name assignment after template updates.
- Both new tests are well-designed: `test_object_names_follow_template_id` validates no cross-contamination across sessions, and `test_template_update_without_object_names_clears_stale_mapping` covers the stale-entry edge case.
- No new issues found in iteration 4.

**Review by banidoru — APPROVED** [review]
- **Approve.** Core bug fix is correct — object names now resolve per-set via `template_id` instead of grabbing the first arbitrary map entry.
- Stale `object_names_map` cleanup on `None` prevents silent wrong-name contamination across template updates.
- `get_template_key()` short-circuit logic is safe given the `retain()`-based mutual exclusion between temporary/applied maps.
- New tests cover both the cross-session isolation and stale-mapping-clear scenarios with proper negative assertions.
- No new issues found in iteration 4.

**Review by banidoru — APPROVED** [review]
All reviewers approved. LGTM.

**Comment by Pterosaur** [comment]
Hi @prsunny , could you please help to merge this PR?

**Review by prsunny — APPROVED** [review]

**Comment by vmittal-msft** [comment]
@prsunny @Pterosaur is this must fix for 202511 ?
